src/include/model_server/model_server_manager.h

```diff
@@ -0,0 +1,127 @@
+
+namespace noisepage::model {
+static constexpr const char *MODEL_CONN_ID_NAME = "model-server-conn";
+static constexpr const char *MODEL_TARGET_NAME = "model";
+static constexpr const char *MODEL_IPC_PATH = "/tmp/model-server-ipc";
+static constexpr const char *MODEL_TCP_HOST = "127.0.0.1";
+static constexpr const int MODEL_TCP_PORT = 15645;
+static constexpr const int INVALID_PID = 0;
+
+/*
+ * Use 128 as convention to indicate failure in a subprocess:
+ * https://www.gnu.org/software/libc/manual/html_node/Exit-Status.html
+ */
+static constexpr const unsigned char MODEL_ERROR_BINARY = 128;
+
+/**
+ * Inteface for pilot related operations
+ */
+class ModelServerManager {
+ public:
+  /**
+   * Construct a ModelServerManager with the given executable script to the python ModelServer
+   * @param model_bin python script path
+   * @param messenger Messenger pointer
+   */
+  ModelServerManager(const std::string &model_bin, const common::ManagedPointer<messenger::Messenger> &messenger);
+
+  /**
+   * Stop the python ModelServer when exits
+   */
+  ~ModelServerManager() { StopModelServer(); }
+
+  /**
+   * Stop the model-pilot daemon by sending a message to the Python model server
+   */
+  void StopModelServer();
+
+  /**
+   * Check if the model server has started.
+   * The user of this function should poll this until the pid changes.
+   * It is async because listening on the messenger's connector is now asynchronous.
+   *
+   * A true return value doesn't mean the model server script is running, but only guarantees
+   * the fork() has succeeded.
+   *
+   * @return true if model server has started
+   */
+  bool ModelServerStarted() const { return py_pid_ != INVALID_PID; }
+
+  /**
+   * Get the python model-server's PID
+   * @return  pid
+   */
+  pid_t GetModelPid() const { return py_pid_; }
+
+  /*******************************************************
+   * ModelServer <-> ModelServerManager logic routines
+   *******************************************************/
+
+  /**
+   * Ask the model server to print a message
+   * This function does not expect a callback
+   *
+   * @warning The message should not contain '-' as per current protocol
+   * @param msg Message to print
+   */
+  void PrintMessage(const std::string &msg);
+
+  /**
+   * Train a model with the given seq files directory
+   *
+   * This function should expect a path to the saved model map
+   * @param model_name Name of the model to train
+   * @param seq_files_dir Seq files's enclosing directory (seq files generated by mini_runner)
+   * @bug Currently NOOP for callback
+   */
+  void TrainWith(const std::string &model_name, const std::string &seq_files_dir);
+
+  /**
+   * Perform inference on the given data file.
+   *
+   * This function should expect some return (Not figured out yet)
+   * @param data_file Seq file
+   * @param model_map_path Model map (should probably be obtained from TrainWith call)
+   * @bug What to do with return and callback?
```

Comment:
We do need to get the predictions back... what's the convention to get the results back in our message framework?